File: AgentforceSDK-ReactNative-Bridge/android/src/main/java/com/salesforce/android/reactagentforce/BridgeUIDelegate.kt
/*
 * Copyright (c) 2024-present, salesforce.com, inc.
 * All rights reserved.
 *
 * Bridge UIDelegate that forwards Agentforce SDK UI events to React Native JS via events.
 */
package com.salesforce.android.reactagentforce

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.salesforce.android.agentforcesdkimpl.AgentConversation
import com.salesforce.android.agentforcesdkimpl.AgentforceConversation
import com.salesforce.android.agentforcesdkimpl.AgentforceUIDelegate
import com.salesforce.android.agentforceservice.AgentforceUtterance
import com.salesforce.android.agentforceservice.conversationservice.data.AgentforceMessage
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.withTimeout
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Implements the Agentforce SDK AgentforceUIDelegate interface and forwards UI events
 * to JavaScript via NativeEventEmitter events.
 *
 * UI event forwarding is disabled by default and enabled when JS registers a UIDelegate.
 */
class BridgeUIDelegate(private val reactContext: ReactContext) : AgentforceUIDelegate {

    /** Controls whether UI events are forwarded to JS. */
    @Volatile
    var forwardingEnabled: Boolean = false

    /** Pending modify-utterance requests awaiting JS responses. */
    private val pendingModifications = ConcurrentHashMap<String, CompletableDeferred<String>>()

    /** Timeout for awaiting a modified utterance from JavaScript (milliseconds). */
    private val modifyUtteranceTimeoutMs: Long = 10_000

    private fun isoTimestamp(): String =
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }.format(Date())

    /** Returns a conversation ID string from an AgentConversation. */
    private fun getConversationId(conversation: AgentConversation): String {
        return if (conversation is AgentforceConversation) {
            conversation.id
        } else {
            conversation.hashCode().toString()
        }
    }

    override suspend fun modifyUtteranceBeforeSending(
        agentforceUtterance: AgentforceUtterance
    ): AgentforceUtterance {
        if (!forwardingEnabled) return agentforceUtterance

        val requestId = UUID.randomUUID().toString()
        val deferred = CompletableDeferred<String>()
        pendingModifications[requestId] = deferred

        try {
            val params = Arguments.createMap().apply {
                putString("requestId", requestId)
                putString("utterance", agentforceUtterance.utterance)
            }

            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("onModifyUtteranceRequest", params)

            val modifiedText = withTimeout(modifyUtteranceTimeoutMs) { deferred.await() }

            return AgentforceUtterance(
                utterance = modifiedText,
                agentforceAttachment = agentforceUtterance.agentforceAttachment
            )
        } catch (e: TimeoutCancellationException) {
            Log.d(TAG, "modifyUtterance timed out for request $requestId, using original")
            return agentforceUtterance
        } catch (e: Exception) {
            Log.w(TAG, "modifyUtterance failed for request $requestId", e)
            return agentforceUtterance
        } finally {
            pendingModifications.remove(requestId)
        }
    }

    /**
     * Called by AgentforceModule when JS responds to a modify-utterance request.
     *
     * @return true if the request was found and completed, false otherwise.
     */
    fun completeModification(requestId: String, modifiedUtterance: String): Boolean {
        val deferred = pendingModifications[requestId] ?: return false
        return deferred.complete(modifiedUtterance)
    }

    override fun didSendUtterance(agentforceUtterance: AgentforceUtterance) {
        if (!forwardingEnabled) return

        val params = Arguments.createMap().apply {
            putString("utterance", agentforceUtterance.utterance)
            putBoolean("hasAttachment", agentforceUtterance.agentforceAttachment != null)
            putString("timestamp", isoTimestamp())
        }

        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("onUtteranceSent", params)
    }

    override fun userDidSwitchAgents(newConversation: AgentConversation) {
        if (!forwardingEnabled) return

        val params = Arguments.createMap().apply {
            putString("conversationId", getConversationId(newConversation))
            putString("timestamp", isoTimestamp())
        }

        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("onAgentSwitch", params)
    }

    override fun didReceiveResponse(
        agentforceMessage: AgentforceMessage,
        conversation: AgentConversation
    ) {
        if (!forwardingEnabled) return

        val params = Arguments.createMap().apply {
            putString("responseId", agentforceMessage.id)
            putString("message", agentforceMessage.message ?: agentforceMessage.text)
            putString("type", agentforceMessage.type ?: "agent")
            putString("conversationId", getConversationId(conversation))
            putString("timestamp", formatTimestamp(agentforceMessage.timeStamp))
        }

        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("onAgentResponse", params)
    }

    private fun formatTimestamp(epochMillis: Long): String =
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }.format(Date(epochMillis))

    companion object {
        private const val TAG = "BridgeUIDelegate"
    }
}

File: AgentforceSDK-ReactNative-Bridge/ios/Agentforce/BridgeUIDelegate.swift
/*
 * Copyright (c) 2024-present, salesforce.com, inc. All rights reserved.
 *
 * React Native bridge UI delegate for Agentforce SDK
 */

import Foundation
import SwiftUI
import AgentforceSDK
import AgentforceService

/// Bridge UI delegate that forwards Agentforce SDK UI events to React Native JavaScript
///
/// Implements the iOS Agentforce SDK's AgentforceUIDelegate protocol and emits events
/// to the React Native event emitter. Forwarding can be enabled/disabled
/// to avoid bridge overhead when no JS delegate is registered.
///
/// The `modifyUtteranceBeforeSending` method emits a request to JS and waits
/// (with a timeout) for JS to respond via `awaitModifiedUtterance` on the module.
class BridgeUIDelegate: AgentforceUIDelegate {

    // MARK: - Properties

    /// Weak reference to the module to avoid retain cycles
    weak var module: AgentforceModule?

    /// Splash screen provider that hosts a React Native welcome screen. Held
    /// strongly: the module owns this delegate, and the provider owns no reference
    /// back to it, so there is no retain cycle.
    private let splashScreenProvider: BridgeSplashScreenProvider

    private let lock = NSLock()
    private var _forwardingEnabled: Bool = false

    /// Controls whether UI delegate events should be forwarded to JavaScript.
    /// Thread-safe: written on the JS thread, read on SDK callback threads.
    var forwardingEnabled: Bool {
        get { lock.withLock { _forwardingEnabled } }
        set { lock.withLock { _forwardingEnabled = newValue } }
    }

    /// Timeout for awaiting a modified utterance from JavaScript (seconds).
    /// If JS does not respond within this window, the original utterance is returned.
    var modifyUtteranceTimeout: TimeInterval = 10.0

    // MARK: - Initialization

    init(module: AgentforceModule, splashScreenProvider: BridgeSplashScreenProvider) {
        self.module = module
        self.splashScreenProvider = splashScreenProvider
    }

    // MARK: - AgentforceUIDelegate Protocol Implementation

    func modifyUtteranceBeforeSending(_ utterance: AgentforceUtterance) async -> AgentforceUtterance {
        guard forwardingEnabled, let module = module else {
            return utterance
        }

        let requestId = UUID().uuidString

        // Ask JS to modify the utterance; the module handles the continuation/timeout internally.
        let modifiedText = await module.awaitModifiedUtterance(
            requestId: requestId,
            utteranceText: utterance.utterance,
            timeout: modifyUtteranceTimeout
        )

        if let modifiedText = modifiedText {
            return AgentforceUtterance(utterance: modifiedText, attachment: utterance.attachment)
        }

        return utterance
    }

    func didSendUtterance(_ utterance: AgentforceUtterance) {
        guard forwardingEnabled else { return }

        let payload: [String: Any] = [
            "utterance": utterance.utterance,
            "hasAttachment": utterance.attachment != nil,
            "timestamp": ISO8601DateFormatter().string(from: Date()),
        ]

        module?.emitUtteranceSentEvent(payload)
    }

    func userDidSwitchAgents(newConversation: any AgentConversation) {
        guard forwardingEnabled else { return }

        let payload: [String: Any] = [
            "conversationId": newConversation.conversationId.uuidString,
            "timestamp": ISO8601DateFormatter().string(from: Date()),
        ]

        module?.emitAgentSwitchEvent(payload)
    }

    func userInitiatedVoice(for conversation: any AgentConversation) {
        // Voice is managed internally in 260.5 - no forwarding needed.
    }

    func didReceiveResponse(_ message: AgentforceMessage, from conversation: any AgentConversation) {
        guard forwardingEnabled, message.state == .finished else { return }

        // AgentforceSDK.AgentforceMessage does not yet expose `id` or `type`.
        // Generate a local responseId and derive type from isUserMessage until
        // the SDK surfaces those properties.
        let payload: [String: Any] = [
            "responseId": UUID().uuidString,
            "message": message.message ?? NSNull(),
            "type": message.isUserMessage ? "user" : "agent",
            "conversationId": conversation.conversationId.uuidString,
            "timestamp": ISO8601DateFormatter().string(from: Date()),
        ]

        module?.emitAgentResponseEvent(payload)
    }

    // Splash screen is not gated by `forwardingEnabled`: it renders a registered
    // React Native component (independent of the JS event stream), and the provider
    // returns nil when no component is registered for the agent.
    @MainActor
    func splashScreen(
        forAgent agentId: String,
        utteranceDelegate: AgentforceSplashScreenUtteranceDelegate
    ) -> AnyView? {
        splashScreenProvider.splashScreen(forAgent: agentId, utteranceDelegate: utteranceDelegate)
    }
}
